### ISSUE TITLE
No Docker client strategy found

### ISSUE THREAD

**Issue by feg624** [testcontainers/testcontainers-node#359]
**Expected Behaviour**
Testcontainers to be able to start a docker-compose environment without problem.

**Actual Behaviour**
It shows the error "No Docker client strategy found".

**Testcontainer Logs**
    No Docker client strategy found

      at ../node_modules/testcontainers/dist/docker/docker-client.js:46:11
      at fulfilled (../node_modules/testcontainers/dist/docker/docker-client.js:5:58)

**Steps to Reproduce**
1. I have manually installed docker on WSL2, so the environment is not standard. However, my execution runs okay up to version 8.6.0 of testcontainers, and it starts failing with this error from version 8.6.1.
2. Due to my installation within WSL2, I need to use the variables TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/mnt/wsl/shared-docker/docker.sock TESTCONTAINERS_RYUK_PRIVILEGED=true
3. Run npm run test
4. See error (shown above)

**Environment Information**
- Operating System: WSL2
- Docker Version: 20.10.13
- Node version: v16.14.1
- Testcontainers version: 8.6.1 and above

Additional info: I've recently started working with testcontainers (which is excellent BTW), and I started using the latest version. As I got the error described above, and after reading some other issues on the Github project, I tried downgrading the version of testcontainers, and it runs okay up to version 8.6.0, and it starts failing from version 8.6.1 and above.

**Comment by cristianrgreco** [comment]
Hi @feg624, please share all the testcontainers logs.

**Comment by feg624** [comment]
Hi @cristianrgreco please check whether the following are good enough. I've got these with version 8.6.1:

DEBUG=testcontainers* TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/mnt/wsl/shared-docker/docker.sock TESTCONTAINERS_RYUK_PRIVILEGED=true npm run test

> my-project@1.0.0 test
> jest

  testcontainers DEBUG Found applicable Docker client strategy: ConfigurationStrategy +0ms
  testcontainers DEBUG Testing Docker client strategy URI: tcp://unix:///mnt/wsl/shared-docker/docker.sock +64ms
  testcontainers INFO  Starting DockerCompose environment: testcontainers-69d6155fe7182257477193f9f34a837d +20ms
  testcontainers DEBUG Creating new Reaper for session: a983c52bf60a8f13139ac6dbbfa672c0 +0ms
  testcontainers DEBUG No registry auth locator found for registry: "https://index.docker.io/v1/" +22ms
  testcontainers WARN  Docker daemon is not reachable: Error: getaddrinfo ENOTFOUND unix +2s
  testcontainers WARN  Docker client strategy ConfigurationStrategy is not reachable +0ms
  testcontainers DEBUG Found applicable Docker client strategy: UnixSocketStrategy +1ms
  testcontainers DEBUG Testing Docker client strategy URI: unix:///var/run/docker.sock +0ms
  testcontainers WARN  Docker daemon is not reachable: Error: connect EACCES /var/run/docker.sock +10ms
  testcontainers WARN  Docker client strategy UnixSocketStrategy is not reachable +0ms
  testcontainers ERROR Failed to list images: Error: No Docker client strategy found +1ms
  testcontainers ERROR Failed to pull image "testcontainers/ryuk:0.3.2": Error: No Docker client strategy found +0ms
 FAIL  test/basic.test.js
  ● DockerComposeEnvironment › works

    No Docker client strategy found

      at ../node_modules/testcontainers/dist/docker/docker-client.js:46:11
      at fulfilled (../node_modules/testcontainers/dist/docker/docker-client.js:5:58)


  ● Test suite failed to run


When I run the same with version 8.6.0, I get the following (success):

DEBUG=testcontainers* TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/mnt/wsl/shared-docker/docker.sock TESTCONTAINERS_RYUK_PRIVILEGED=true npm run test

> my-project@1.0.0 test
> jest

  testcontainers DEBUG Found applicable Docker client strategy: ConfigurationStrategy +0ms
  testcontainers DEBUG Testing Docker client strategy URI: unix:///mnt/wsl/shared-docker/docker.sock +1ms
  testcontainers INFO  Starting DockerCompose environment: testcontainers-896c8f46d9e81b03e0377dc4ad444d36 +82ms
  testcontainers DEBUG Creating new Reaper for session: 5b8833e09acd5a09cb66316de4489371 +1ms
  testcontainers DEBUG No registry auth locator found for registry: "https://index.docker.io/v1/" +13ms
  testcontainers INFO  Using Docker client strategy: ConfigurationStrategy, Docker host: localhost +68ms
  testcontainers DEBUG Not pulling image as it already exists: testcontainers/ryuk:0.3.2 +465ms
  testcontainers INFO  Creating container for image: testcontainers/ryuk:0.3.2 +1ms
  testcontainers INFO  Starting container testcontainers/ryuk:0.3.2 with ID: fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +205ms
  testcontainers DEBUG System diagnostics: {"node":{"version":"v16.14.1","architecture":"x64","platform":"linux"},"docker":{"serverVersion":"20.10.13","operatingSystem":"Ubuntu 20.04.4 LTS","operatingSystemType":"linux","architecture":"x86_64","cpus":4,"memory":6529802240},"dockerCompose":{"version":"1.25.0"}} +2s
  testcontainers DEBUG Waiting for container to be ready: fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +456ms
  testcontainers DEBUG Waiting for host port 49196 for fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +0ms
  testcontainers DEBUG Waiting for internal port 8080 for fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +2ms
  testcontainers DEBUG Host port 49196 ready for fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +6ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 Pinging Docker... +0ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 Docker daemon is available! +14ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 Starting on port 8080... +6ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 Started! +29ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 New client connected: 127.0.0.1:34853 +183ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 EOF +0ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 Client disconnected: 127.0.0.1:34853 +4ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:43 Received the first connection +0ms
  testcontainers DEBUG Internal port 8080 ready for fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +430ms
  testcontainers INFO  Container is ready +0ms
  testcontainers DEBUG Connecting to Reaper fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 on localhost:49196 +1ms
  testcontainers DEBUG Connected to Reaper fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24 +2ms
  testcontainers INFO  Upping DockerCompose environment +1ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:44 New client connected: 172.17.0.1:53286 +107ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:44 Adding {"label":{"org.testcontainers.session-id=5b8833e09acd5a09cb66316de4489371":true}} +5ms
  testcontainers:containers TRACE fcf6a1f7d4cdf7ab2402e3f5b798813f0b3cf607698ef452a34fd94abf38cf24: 2022/04/27 13:26:44 Adding {"label":{"com.docker.compose.project=testcontainers-896c8f46d9e81b03e0377dc4ad444d36":true}} +1ms
  testcontainers INFO  Upped DockerCompose environment +10s
  testcontainers INFO  Started the following containers: /swagger, /nginx, /rest-api, /flyway, /cube-api, /postgres +34ms

**Comment by cristianrgreco** [comment]
Thanks @feg624, that's very helpful. I see an immediate issue in this log line `testcontainers DEBUG Testing Docker client strategy URI: tcp://unix:///mnt/wsl/shared-docker/docker.sock +64ms` but the solution at the moment is unclear.

Out of curiosity, the `ConfigurationStrategy` is only applicable if the `DOCKER_HOST` environment variable is set, is this the case for you?

**Comment by feg624** [comment]
Thanks @cristianrgreco for the quick check on your side. As I mentioned earlier, I made a "custom" docker install within WSL2, so it's not installed via docker desktop on Windows.

The docker related environment variables are:

DOCKER_BUILDKIT=1
DOCKER_DIR=/mnt/wsl/shared-docker
DOCKER_DISTRO=Ubuntu
DOCKER_HOST=unix:///mnt/wsl/shared-docker/docker.sock
DOCKER_SOCK=/mnt/wsl/shared-docker/docker.sock

**Comment by cristianrgreco** [comment]
OK, I think I understand why this is so confusing 😄 

Because the `DOCKER_HOST` environment variable is set, we use the `ConfigurationStrategy` to create the Docker client. The `TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE` is irrelevant here, it is later used by the resource reaper, whereas in this issue it is failing to even initialise the Docker client. 

In version 8.6.0, we parse the `DOCKER_HOST` to get the hostname and port. Testing it now, the result of `const { hostname, port } = new URL('unix:///mnt/wsl/shared-docker/docker.sock');` is empty strings. The underlying `docker-modem` will set the `socketPath` to the value of `DOCKER_HOST` by default (https://github.com/apocas/docker-modem/blob/8556ad38cf5a022b22f50920fae6b08d2ac3f207/lib/modem.js#L23-L30), and this is not overridden by the `ConfigurationStrategy` which only sets the host and port https://github.com/testcontainers/testcontainers-node/blob/bcfadd26de23e09e1b6056a618a67f51d9957a59/src/docker/docker-client.ts#L67-L70

The `host` passed in from testcontainers is parsed by the modem (https://github.com/apocas/docker-modem/blob/8556ad38cf5a022b22f50920fae6b08d2ac3f207/lib/modem.js#L37), which in this case has undefined `hostname` (`tcp://`). The modem later checks that the host is not set and sets the socket as before (https://github.com/apocas/docker-modem/blob/8556ad38cf5a022b22f50920fae6b08d2ac3f207/lib/modem.js#L80).

Basically 8.6.0 worked by accident here. It incorrectly parsed the host/port from the `DOCKER_HOST`, but the underlying modem detected and use the socket path anyway.

I will think about how to fix this. Perhaps the `ConfigurationStrategy` needs to check that if it is a valid URL, then set the `host`/`port` accordingly, else set the `socketPath` directly to the value of `DOCKER_HOST`. 

I will release a beta once implemented.

**Comment by feg624** [comment]
Great! Send me the info and I'll test it out.

**Comment by cristianrgreco** [comment]
@feg624 could you please try version `8.8.1-beta.0`?

**Comment by feg624** [comment]
Hi @cristianrgreco , I've just tested the beta, but similar error as before. Please check the logs below:

DEBUG=testcontainers* TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/mnt/wsl/shared-docker/docker.sock TESTCONTAINERS_RYUK_PRIVILEGED=true npm run test

> my-project@1.0.0 test
> jest

  testcontainers DEBUG Found applicable Docker client strategy: ConfigurationStrategy +0ms
  testcontainers DEBUG Testing Docker client strategy URI: unix:///mnt/wsl/shared-docker/docker.sock +58ms
  testcontainers INFO  Starting DockerCompose environment: testcontainers-04e7f7d5f99427833a8094fa3128ded7 +13ms
  testcontainers DEBUG Creating new Reaper for session: c88613f7655c786e125181d14acb638c +0ms
  testcontainers WARN  Docker daemon is not reachable: Error: connect ENOENT unix:///mnt/wsl/shared-docker/docker.sock +10ms
  testcontainers WARN  Docker client strategy ConfigurationStrategy is not reachable +1ms
  testcontainers DEBUG Found applicable Docker client strategy: UnixSocketStrategy +0ms
  testcontainers DEBUG Testing Docker client strategy URI: unix:///var/run/docker.sock +1ms
  testcontainers WARN  Docker daemon is not reachable: Error: connect EACCES /var/run/docker.sock +13ms
  testcontainers WARN  Docker client strategy UnixSocketStrategy is not reachable +0ms
  testcontainers DEBUG No registry auth locator found for registry: "https://index.docker.io/v1/" +6ms
  testcontainers ERROR Failed to list images: Error: No Docker client strategy found +2ms
  testcontainers ERROR Failed to pull image "testcontainers/ryuk:0.3.2": Error: No Docker client strategy found +0ms
(node:5538) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 1)
(Use `node --trace-warnings ...` to show where the warning was created)
 FAIL  test/basic.test.js
  ● DockerComposeEnvironment › Instantiates a building block approval process

    No Docker client strategy found

      at ../node_modules/testcontainers/dist/docker/docker-client.js:46:11
      at fulfilled (../node_modules/testcontainers/dist/docker/docker-client.js:5:58)

I also tried removing the TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE environment variable, but same result.

**Comment by cristianrgreco** [comment]
@feg624 Please try `8.8.1-beta.1`

**Comment by feg624** [comment]
Excellent @cristianrgreco , version `8.8.1-beta.1` worked great! And you're right, it only worked by defining the env variable `TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE`.

Thanks a lot for your quick support on this!

**Comment by cristianrgreco** [comment]
Great, released in [8.9.0](https://github.com/testcontainers/testcontainers-node/releases/tag/v8.9.0). Thanks for raising the issue and for your help getting it fixed! All the best
